File: example/kotlinlib/script/10-static-site/StaticSite.kt
//| mvnDeps:
//| - org.jetbrains.kotlinx:kotlinx-html-jvm:0.11.0
//| - com.atlassian.commonmark:commonmark:0.13.1

import java.io.File
import org.commonmark.parser.Parser
import org.commonmark.renderer.html.HtmlRenderer
import kotlinx.html.*
import kotlinx.html.stream.appendHTML

fun main(args: Array<String>) {
    val outDir = File("site-out")
    val outPostDir = outDir.resolve("post")

    val postInfo = File("post").listFiles { f -> f.extension == "md" }!!
        .map { f ->
            val match = Regex("""(\d+)\s*-\s*(.+)\.md""").matchEntire(f.name)
                ?: error("Invalid post filename: ${f.name}")
            val (prefix, suffix) = match.destructured
            Triple(prefix.toInt(), suffix, f)
        }
        .sortedBy { it.first }

    if (outDir.exists()) outDir.deleteRecursively()
    outPostDir.mkdirs()

    fun writeHtml(path: File, pageTitle: String, bodyContent: FlowContent.() -> Unit) {
        path.writer().use { w ->
            w.appendLine("<!DOCTYPE html>")
            w.appendHTML().html {
                body {
                    h1 { +pageTitle }
                    bodyContent()
                }
            }
        }
    }

    val parser = Parser.builder().build()
    val renderer = HtmlRenderer.builder().build()

    for ((_, suffix, file) in postInfo) {
        val document = parser.parse(file.readText())
        val htmlContent = renderer.render(document)

        val slug = suffix.replace(" ", "-").lowercase()
        val outputFile = outPostDir.resolve("$slug.html")

        writeHtml(outputFile, "Blog / $suffix") {
            div{
                unsafe {
                    +htmlContent
                }
            }
        }
    }

    val indexFile = outDir.resolve("index.html")
    writeHtml(indexFile, "Blog") {
        for ((_, suffix, _) in postInfo) {  h2 { +suffix }  }
    }
}

File: example/scalalib/script/5-static-site/StaticSite.scala
//| mvnDeps:
//| - com.lihaoyi::scalatags:0.13.1
//| - com.atlassian.commonmark:commonmark:0.13.1
import scalatags.Text.all._

@main def main() = {
  val postInfo = os
    .list(os.pwd / "post")
    .map { p =>
      val s"$prefix - $suffix.md" = p.last
      (prefix, suffix, p)
    }
    .sortBy(_._1.toInt)

  os.remove.all(os.pwd / "site-out")
  os.makeDir.all(os.pwd / "site-out/post")

  for ((_, suffix, path) <- postInfo) {
    val parser = org.commonmark.parser.Parser.builder().build()
    val document = parser.parse(os.read(path))
    val renderer = org.commonmark.renderer.html.HtmlRenderer.builder().build()
    val output = renderer.render(document)
    os.write(
      os.pwd / "site-out/post" / (suffix.replace(" ", "-").toLowerCase + ".html"),
      doctype("html")(
        html(
          body(
            h1(a("Blog"), " / ", suffix),
            raw(output)
          )
        )
      )
    )
  }
}
